commit 0dc96c73bc31b75708856e27c142329a1726c038 (refs/changes/58/2858/4)
Author: Rui Loura <rui@joyent.com>
Date:   2017-11-01T15:06:40-04:00 (1 year, 11 months ago)
    
    NET-377 sdc-nat zones create poor ipnat.conf configuration
    Reviewed by: Cody Mello <cody.mello@joyent.com>
    Approved by: Cody Mello <cody.mello@joyent.com>

diff --git a/bin/setup-nat b/bin/setup-nat
index 0cdd836..472dd34 100755
--- a/bin/setup-nat
+++ b/bin/setup-nat
@@ -16,6 +16,8 @@
 set -o xtrace
 set -o pipefail
 
+NATCONF=/etc/ipf/ipnat.conf
+
 function fatal
 {
     local msg="$*"
@@ -35,7 +37,7 @@ function enable_ipnat
 {
     /usr/sbin/ipf -E || fatal "failed to enable ipf"
     /usr/sbin/ipnat -CF || fatal "failed to flush ipnat rules"
-    /usr/sbin/ipnat -f /etc/ipf/ipnat.conf || fatal "failed to flush ipnat rules"
+    /usr/sbin/ipnat -f $NATCONF || fatal "failed to set ipnat rules"
 }
 
 #
@@ -60,15 +62,23 @@ function write_config
 
     first=
     for nic in $primary; do
-            [[ -n "$first" ]] && fatal "encountered multiple primary nics"
-            first="fatal"
+        [[ -n "$first" ]] && fatal "encountered multiple primary nics"
+        first="fatal"
     done
 
     mkdir -p /etc/ipf || fatal "failed to mkdir /etc/ipf"
-    echo "map $primary $subnet -> 0/32" > /etc/ipf/ipnat.conf || \
-        fatal "failed to write out ipnat.conf"
+
+    #
+    # We need 2 rules here.  One for TCP & UDP that will translate both the IP
+    # address and the port, and one for everything else (e.g. ICMP) that will
+    # only translate the IP address.
+    #
+    cat << EOF > $NATCONF || fatal "failed to write $NATCONF"
+map $primary $subnet -> 0/32 portmap tcp/udp 1025:65535
+map $primary $subnet -> 0/32
+EOF
 }
 
 enable_ipv4_forwarding || fatal "failed to enable routing"
-write_config || fatal "failed to write ipnat config"
-enable_ipnat || fatal "failed to enable ipneat"
+write_config || fatal "failed to write $NATCONF"
+enable_ipnat || fatal "failed to enable ipnat"
